apps/web/src/dialogs/command-palette/command-palette-dialog.tsx

```diff
@@ -0,0 +1,461 @@
+import { CustomScrollbarsVirtualList } from "../../components/list-container";
+import { hashNavigate, navigate } from "../../navigation";
+import { useEditorStore } from "../../stores/editor-store";
+import Config from "../../utils/config";
+import { commands as COMMANDS } from "./commands";
+
+interface Command {
+  id: string;
+  title: string;
+  highlightedTitle?: string;
+  type: "command" | "note" | "notebook" | "tag" | "reminder";
+  group: string;
+}
+
+type GroupedCommands = { group: Command["group"]; count: number }[];
+
+export const CommandPaletteDialog = DialogManager.register(
+  function CommandPaletteDialog(props: BaseDialogProps<boolean>) {
+    const [commands, setCommands] = useState<Command[]>(getDefaultCommands());
+    const [selected, setSelected] = useState(0);
+    const [query, setQuery] = useState(">");
+    const [loading, setLoading] = useState(false);
+    const virtuosoRef = useRef<GroupedVirtuosoHandle>(null);
+
+    useEffect(() => {
+      virtuosoRef.current?.scrollToIndex({
+        index: selected,
+        align: "end",
+        behavior: "auto"
+      });
+    }, [selected]);
+
+    async function onChange(e: React.ChangeEvent<HTMLInputElement>) {
+      setSelected(0);
+      const query = e.target.value;
+      setQuery(query);
+      !isCommandMode(query) && setLoading(true);
+      const res = await search(query);
+      const highlighted = db.lookup.fuzzy(
+        prepareQuery(query),
+        res.map((r) => ({
+          ...r,
+          highlightedTitle: r.title
+        })) ?? [],
+        "highlightedTitle",
+        {
+          prefix: "<b style='color: var(--accent-foreground)'>",
+          suffix: "</b>"
+        }
+      );
+      setCommands(highlighted ?? []);
+      setLoading(false);
+    }
+
+    const grouped = commands.reduce((acc, command) => {
+      const item = acc.find((c) => c.group === command.group);
+      if (item) {
+        item.count++;
+      } else {
+        acc.push({ group: command.group, count: 1 });
+      }
+      return acc;
+    }, [] as GroupedCommands);
+
+    return (
+      <Dialog
+        isOpen={true}
+        width={650}
+        onClose={() => {
+          props.onClose(false);
+        }}
+        noScroll
+        sx={{
+          fontFamily: "body"
+        }}
+      >
+        <Box
+          className="ping"
+          sx={{
+            height: 4,
+            bg: loading ? "accent" : "background",
+            transition: "background 0.2s"
+          }}
+        />
+        <Flex
+          variant="columnFill"
+          sx={{ mx: 3, overflow: "hidden", height: 400 }}
+          onKeyDown={(e) => {
+            if (e.key == "Enter") {
+              e.preventDefault();
+              const command = commands[selected];
+              if (!command) return;
+              const action = getCommandAction({
+                id: command.id,
+                type: command.type
+              });
+              if (action) {
```

Comment:
Why wouldn't there be an action for a command?

Reply:
Because... TypeScript 😆 

`getCommandAction` has a possibility of returning `undefined` because of `CommandActionMap.get`.  Do you think we should use a non-null asssertion operator (`!`) when we `get` from the map? For now, I replaced `if (action)` with `action?.`